sisl/geom/bilayer.py

```diff
@@ -0,0 +1,116 @@
+    Laissardiere et al., Nano Lett. 10, 804-808 (2010).
+
+
+    Parameters
+    ----------
+    bond : 1.42
+       bond length (in Angstrom) between atoms in the honeycomb lattice
+    bottom_atom : Atom(6)
+       atom (or atoms) in the bottom layer
+    top_atom : Atom(6)
+       atom (or atoms) in the top layer
+    stacking : {'AB', 'AA'}
+       stacking sequence of the bilayer
+    twist : (0, 0)
+       integer coordinates (m, n) defining a commensurate twist angle
+    separation : float
+       distance between the two layers (in Angstrom)
+    ret_angle : bool
+       return the twist angle (in degrees) in addition to the geometry instance
+    layer : {'both', 'bottom', 'top'}
+       control which layer(s) to include in geometry
+    """
+    if bottom_atom is None:
+        bottom_atom = Atom(Z=6, R=bond * 1.01)
+    if top_atom is None:
+        top_atom = Atom(Z=6, R=bond * 1.01)
+
+    # Construct two layers
+    bottom = geom.honeycomb(bond=bond, atom=bottom_atom)
+    top = geom.honeycomb(bond=bond, atom=top_atom)
+    honeycomb = bottom.copy()
+
+    if stacking.lower() == 'aa':
+        top = top.move([0, 0, separation])
+    elif stacking.lower() == 'ab':
+        top = top.move([bond, 0, separation])
+
+    # Compute twist angle
+    m, n = twist
+    m, n = abs(m), abs(n)
+    if m > n:
+        # Set m as the smaller integer
+        m, n = n, m
+
+    if not (isinstance(n, int) and isinstance(m, int)):
+        raise RuntimeError('twist coordinates need to be integers!')
+
+    if m == n:
+        # No twisting
+        theta = 0
+        rep = 1
+        natoms = 2
+    else:
+        # Twisting
+        cos_theta = (n ** 2 + 4 * n * m + m ** 2) / (2 * (n ** 2 + n * m + m ** 2) )
+        theta = np.arccos(cos_theta) * 360 / (2 * np.pi)
+        rep = 4 * (n + m)
+        natoms = 2 * (n ** 2 + n * m + m ** 2)
+
+    if rep > 1:
+        # Set origo through an A atom near the middle of the geometry
+        bottom = bottom.tile(rep, axis=0).tile(rep, axis=1)
+        top = top.tile(rep, axis=0).tile(rep, axis=1)
+        tvec = rep * (honeycomb.cell[0] + honeycomb.cell[1]) / 2
+        bottom = bottom.move(-tvec)
+        top = top.move(-tvec)
+
+        # Set new lattice vectors
+        bottom.cell[0] = n * honeycomb.cell[0] + m * honeycomb.cell[1]
+        bottom.cell[1] = -m * honeycomb.cell[0] + (n + m) * honeycomb.cell[1]
+
+        # Rotate top layer around A atom in bottom layer
+        top = top.rotate(theta, [0, 0, 1])
+
+        top.cell[:] = bottom.cell[:]
+
+    # Which layers to be returned
+    if layer.lower() == 'bottom':
+        bilayer = bottom
+    elif layer.lower() == 'top':
+        bilayer = top
+    else:
+        bilayer = bottom.add(top)
+        natoms *= 2
+
+    if rep > 1:
+        # Remove atoms outside cell
+        center = 1e-3 * np.array([1, 1, 1])
+        cell_box = Cuboid(1.0 * bilayer.cell[:], center=center)
```

Comment:
I'll have a quick look at this. I think for clarity `center` should be shifted slightly negatively (to be consistent with the first atom of the bottom layer.

Reply:
Maybe it would be better instead to select the atoms in each layer in the final cell before joining the two into one geometry? In order to have all bottom atoms appearing before the top ones (not sure if this ordering is currently respected)?

Reply:
I am pretty sure this is respected. The indices are in ascending order, so first bottom, then top.
I guess for extremely big systems it would be better to first create bottom, remove atoms, then create top, then remove atoms, then add.
But I don't think it is too important, the algorithms should be more or less the same. :)